File: 2021/10/14/main.go
package main

func numSquares(n int) int {
	p := []int{}
	for i := 1; i*i <= n; i++ {
		p = append(p, i*i)
	}

	dp := make([]int, n+1)
	for i := 1; i < len(dp); i++ {
		dp[i] = (n + 1) * (n + 1)
	}

	for _, v := range p {
		for i := v; i < len(dp); i++ {
			if dp[i] > dp[i-v]+1 {
				dp[i] = dp[i-v] + 1
			}
		}
	}

	return dp[n]
}

File: 2021/10/14/main.py
class Solution:
    def numSquares(self, n: int) -> int:
        p = [i*i for i in range(1, n+1)]
        dp = [(n + 1) * (n + 1)] * (n+1)
        dp[0] = 0
        for v in p:
            for i in range(v, n+1):
                if dp[i] > dp[i-v]+1:
                    dp[i] = dp[i-v] + 1

        return dp[n]
